implement metrics for grafanaOrganizations (#533)
giantswarm/observability-operator: docs/monitoring/grafanaorganization-monitoring.md

# GrafanaOrganization Monitoring Guide

> This guide explains how to monitor GrafanaOrganization custom resources deployed in your Kubernetes cluster using the observability-operator's built-in metrics and monitoring capabilities.

Reviewer: Who is this doc for ? Maintainers of this project or external users ?

I feel strange that we are talking about general "how to monitor" grafana orgs, while later on we are very specific on the implementations details of how its done here.

Also this doc overlap with the previous one which describe how to **use** the metrics while this doc explains the how to **instrument** the code to produce metrics, concerns should be clearly separated here.

## Overview

The observability-operator exposes prometheus-compatible metrics.

> On top of the common metrics available for all renconcilers, the `grafanaOrganization` reconciler provides metrics per:

Reviewer: ```suggestion
On top of the [common metrics exposed by the controller runtime](https://book.kubebuilder.io/reference/metrics-reference), the `grafanaOrganization` reconciler provides metrics per:
```

- organization
- tenant

## Available Metrics

The operator exposes the following metrics for GrafanaOrganization resources:

### Resource State Metrics

| Metric | Type | Description | Labels |
|--------|------|-------------|--------|
| `observability_operator_grafana_organization_info` | Gauge | Information about GrafanaOrganization resources | `name`, `status` (active, pending, error),`display_name`, `org_id` |

### Configuration Metrics

| Metric | Type | Description | Labels |
|--------|------|-------------|--------|

> | `observability_operator_grafana_organization_tenants` | Gauge | Name of tenants per organization | `name`, `org_id` |

Reviewer: ```suggestion
| `observability_operator_grafana_organization_tenants` | Gauge | Number of tenants per organization | `name`, `org_id` |
```

## Monitoring Best Practices

### Key Metrics to Watch

1. **Resource Status Distribution**
   ```promql
   observability_operator_grafana_organizations_total
   ```
   - Monitor for resources stuck in pending or error states

2. **Reconciliation Success Rate**
   ```promql

>    grafanaorganization:reconciliation_success_rate

Author: Removed this part

   ```
   - Should be >95% under normal conditions
   - Should be <5% under normal conditions

### Custom Queries

#### Tenant Distribution
```promql
topk(10, observability_operator_grafana_organization_tenants)
```

## Extending the code to provide more observability data

### Adding Custom Metrics

To add custom metrics:

1. Define the metric in `pkg/metrics/metrics.go`
2. Update the collector in `pkg/metrics/grafanaorganization_collector.go`
3. Add metric recording in the controller

Example:
```go
// In metrics.go
CustomMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
    Name: "observability_operator_grafana_organization_custom_metric",
    Help: "Custom metric for GrafanaOrganization",
}, []string{"name"})

// In controller
r.metricsCollector.RecordCustomMetric(grafanaOrganization.Name, value)

> ```

Reviewer: Maybe add a link to the upstream doc on how to instrument code https://pkg.go.dev/github.com/prometheus/client_golang@v1.23.2/prometheus

# GrafanaOrganization Monitoring Guide

This guide explains how to monitor GrafanaOrganization custom resources deployed in your Kubernetes cluster using the observability-operator's built-in metrics and monitoring capabilities.

## Overview

> The observability-operator provides comprehensive monitoring for GrafanaOrganization resources through:

Author: I applied your suggestion :)
